src/language_server_protocol.rs

```diff
@@ -3123,10 +3096,66 @@ impl LanguageClient {
             )?;
         }
 
-        info!("End {}", NOTIFICATION__HandleCursorMoved);
         Ok(())
     }
 
+    fn virtual_texts_from_diagnostics(
+        &self,
+        filename: &str,
+        viewport: viewport::Viewport,
+    ) -> Fallible<Vec<VirtualText>> {
+        let mut virtual_texts = vec![];
+        let diagnostics = self.get(|state| state.diagnostics.clone())?;
+        let diagnosticsDisplay = self.get(|state| state.diagnosticsDisplay.clone())?;
+        let diag_list = diagnostics.get(filename);
+        if let Some(diag_list) = diag_list {
+            for diag in diag_list {
+                if viewport.overlaps(diag.range) {
+                    virtual_texts.push(VirtualText {
+                        line: diag.range.start.line,
+                        text: diag.message.replace("\n", "  ").clone(),
+                        hl_group: diagnosticsDisplay
+                            .get(&(diag.severity.unwrap_or(DiagnosticSeverity::Hint) as u64))
+                            .ok_or_else(|| err_msg("Failed to get display"))?
+                            .virtualTexthl
+                            .clone(),
+                    });
+                }
+            }
+        }
+
+        Ok(virtual_texts)
+    }
+
+    fn virtual_texts_from_code_lenses(&self, filename: &str) -> Fallible<Vec<VirtualText>> {
+        let mut virtual_texts = vec![];
+        let code_lenses =
+            self.get(|state| state.code_lens.get(filename).cloned().unwrap_or_default())?;
+
+        for cl in code_lenses {
+            if let Some(command) = cl.command {
+                let line = cl.range.start.line;
+                let text = command.title;
+
+                match virtual_texts
+                    .iter()
+                    .position(|v: &VirtualText| v.line == line)
+                {
+                    Some(idx) => virtual_texts[idx]
+                        .text
+                        .push_str(format!(" | {}", text).as_str()),
+                    None => virtual_texts.push(VirtualText {
```

Comment:
Have you considered `impl From<CodeLens> for VirtualText`?

Reply:
Yeah but it isn't that handy because the virtual text should, at least in the current implementation, be built only if the code lens has a `command` that is not `None`. And also the the text of the final code lens depends on whether other virtual texts are in place in that line, that's why I opted for a standalone function instead.

Reply:
Make sense. Will merge this in.